File: Codes/Level 2/Recursion and Backtracking/Permutations and Combinations/PermutationsLeetCode.java
class Solution {
    
    public void premuteHelper(int[] nums, List<List<Integer>> ans, List<Integer> currAns, boolean[] visited){
        if(currAns.size() == nums.length){
            List<Integer> temp = new ArrayList<>();
            for(Integer i : currAns){
                temp.add(i);
            }
            ans.add(temp);
            return;
        }
        
        for(int i = 0; i<nums.length; i++){
            if(visited[i] == false){
                visited[i] = true;
                currAns.add(nums[i]);
                premuteHelper(nums, ans, currAns, visited);
                visited[i] = false;
                currAns.remove(currAns.size() - 1);
            }
        }
    }
    
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> currAns = new ArrayList<>();
        boolean[] visited = new boolean[nums.length];
        
        premuteHelper(nums, ans, currAns, visited);
        
        return ans;
    }
}

File: Codes/Level 2/Recursion and Backtracking/Permutations and Combinations/PermutationsLeetCode.cpp
class Solution {
public:
    
    void permute(vector<vector<int>> &ans, vector<int> &currAns, vector<int> &nums, vector<bool> &visited){
        if(currAns.size() == nums.size()){
            ans.push_back(currAns);
            return;
        }
        for(int i=0; i<nums.size(); i++){
            if(!visited[i]){
                visited[i] = true;
                currAns.push_back(nums[i]);
                permute(ans, currAns, nums, visited);
                currAns.pop_back();
                visited[i] = false;
            }
        }
    }
    
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> currAns;
        vector<bool> visited(nums.size());
        
        permute(ans, currAns, nums, visited);
        
        return ans;
    }
};